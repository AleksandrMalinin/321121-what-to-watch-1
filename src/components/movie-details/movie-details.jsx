import React, {PureComponent} from 'react';
import {connect} from 'react-redux';
import {Link, withRouter} from "react-router-dom";
import PropTypes from 'prop-types';
import {getFilmId, getFilmsAlike} from '../../reducer/data/selectors.js';
import {getAuthorizationStatus} from '../../reducer/user/selectors.js';
import Sprite from '../sprite/sprite.jsx';
import Logo from '../logo/logo.jsx';
import UserBlock from '../user-block/user-block.jsx';
import Tabs from '../tabs/tabs.jsx';
import MovieList from '../movie-list/movie-list.jsx';
import FullVideoPlayer from '../full-video-player/full-video-player.jsx';
import withFullVideoPlayer from '../../hocs/with-full-video-player/with-full-video-player.js';

class MovieDetails extends PureComponent {
  constructor(props) {
    super(props);

    this.handleFavouriteStatusChange = this._handleFavouriteStatusChange.bind(this);
  }

  render() {
    const {movie, moviesAlike, isPlaying, fullVideoShown, isAuthorizationRequired, onPlayButtonClick} = this.props;
    const moviesAlikeCut = moviesAlike.slice(0, 4);

    if (!movie) {
      return null;
    } else {
      return <React.Fragment>
        <Sprite/>

        <section className="movie-card movie-card--full">
          <div className="movie-card__hero">
            <div className="movie-card__bg">
              <img src={movie.background_image} alt={movie.background_image} />
            </div>

            <h1 className="visually-hidden">WTW</h1>

            <header className="page-header movie-card__head">
              <Logo/>
              <UserBlock/>
            </header>

            <div className="movie-card__wrap">
              <div className="movie-card__desc">
                <h2 className="movie-card__title">{movie.name}</h2>
                <p className="movie-card__meta">
                  <span className="movie-card__genre">{movie.genre}</span>
                  <span className="movie-card__year">{movie.released}</span>
                </p>

                <div className="movie-card__buttons">
                  <button className="btn btn--play movie-card__button" type="button" onClick={onPlayButtonClick}>
                    <svg viewBox="0 0 19 19" width="19" height="19">
                      <use xlinkHref="#play-s"></use>
                    </svg>
                    <span>Play</span>
                  </button>
                  <button className="btn btn--list movie-card__button" type="button" onClick={this.handleFavouriteStatusChange}>
                    {movie && !movie.is_favorite ?
                      <svg viewBox="0 0 19 20" width="19" height="20">
                        <use xlinkHref="#add"></use>
                      </svg> :

                      <svg viewBox="0 0 18 14" width="18" height="14">
                        <use xlinkHref="#in-list"></use>
                      </svg>
                    }
                    <span>My list</span>
                  </button>
                  {!isAuthorizationRequired ?
                    <Link to={`/reviews/add/${movie.id}`} className="btn movie-card__button">Add review</Link>
                    :
                    ``
                  }
                </div>
              </div>
            </div>
          </div>

          <div className="movie-card__wrap movie-card__translate-top">
            <div className="movie-card__info">
              <div className="movie-card__poster movie-card__poster--big">
                <img src={movie.poster_image} alt={movie.poster_image} width="218" height="327" />
              </div>

              <div className="movie-card__desc">
                <Tabs
                  movie={movie}
                />
              </div>

            </div>
          </div>
        </section>

        <div className="page-content">
          <section className="catalog catalog--like-this">
            <h2 className="catalog__title">More like this</h2>

            <MovieList movies={moviesAlikeCut}/>
          </section>

          <footer className="page-footer">
            <div className="logo">
              <Link to="/" className="logo__link logo__link--light">
                <span className="logo__letter logo__letter--1">W</span>
                <span className="logo__letter logo__letter--2">T</span>
                <span className="logo__letter logo__letter--3">W</span>
              </Link>
            </div>

            <div className="copyright">
              <p>Â© 2019 What to watch Ltd.</p>
            </div>
          </footer>
        </div>

        {fullVideoShown ? <FullVideoPlayer onPlayButtonClick={onPlayButtonClick} movie={movie} isPlaying={isPlaying}/> : ``}
      </React.Fragment>;
    }
  }

  _handleFavouriteStatusChange() {
    const id = this.props.movie.id;
    const status = this.props.movie.is_favorite;

    if (this.props.isAuthorizationRequired) {
      return this.props.history.push(`/login`);
    }

    return this.props.onFavouriteStatusChange(id, status);
  }
}

MovieDetails.propTypes = {
  movie: PropTypes.object,
  moviesAlike: PropTypes.array,
  isPlaying: PropTypes.bool,
  fullVideoShown: PropTypes.bool,
  onPlayButtonClick: PropTypes.func,
  isAuthorizationRequired: PropTypes.bool,
  onFavouriteStatusChange: PropTypes.func,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  })
};

const mapStateToProps = (state, ownProps) => ({
  isAuthorizationRequired: getAuthorizationStatus(state),
  movie: getFilmId(state, ownProps.match.params.id),
  moviesAlike: getFilmsAlike(state, ownProps.match.params.id)
});

export {MovieDetails};
export default connect(mapStateToProps)(withRouter(withFullVideoPlayer(MovieDetails)));
