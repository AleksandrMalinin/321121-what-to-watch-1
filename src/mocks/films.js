/* eslint camelcase: ["error", {properties: "never"}] */

export const films = [
  {
    title: `Fantastic Beasts: The Crimes of Grindelwald`,
    preview_image: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    preview_video_link: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Comedies`
  },
  {
    title: `Bohemian Rhapsody`,
    preview_image: `img/bohemian-rhapsody.jpg`,
    preview_video_link: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Crime`
  },
  {
    title: `Macbeth`,
    preview_image: `img/macbeth.jpg`,
    preview_video_link: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Documentary`
  },
  {
    title: `Aviator`,
    preview_image: `img/aviator.jpg`,
    preview_video_link: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Dramas`
  },
  {
    title: `We need to talk about Kevin`,
    preview_image: `img/we-need-to-talk-about-kevin.jpg`,
    preview_video_link: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Horror`
  }
];
