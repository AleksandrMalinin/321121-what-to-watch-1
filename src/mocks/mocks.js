/* eslint camelcase: ["error", {properties: "never"}] */

export const films = [
  {
    name: `Fantastic Beasts: The Crimes of Grindelwald`,
    preview_image: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    preview_video_link: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Comedies`
  },
  {
    name: `Bohemian Rhapsody`,
    preview_image: `img/bohemian-rhapsody.jpg`,
    preview_video_link: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Crime`
  },
  {
    name: `Macbeth`,
    preview_image: `img/macbeth.jpg`,
    preview_video_link: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Documentary`
  },
  {
    name: `Aviator`,
    preview_image: `img/aviator.jpg`,
    preview_video_link: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Dramas`
  },
  {
    name: `We need to talk about Kevin`,
    preview_image: `img/we-need-to-talk-about-kevin.jpg`,
    preview_video_link: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Horror`
  }
];

export const defaultMovie = {
  background_color: `#AD9F8B`,
  background_image: `https://es31-server.appspot.com/wtw/static/film/background/Dardjeeling_Limited.jpg`,
  description: `A year after their father's funeral, three brothers travel across India by train in an attempt to bond with each other.`,
  director: `Wes Anderson`,
  genre: `Adventure`,
  id: 1,
  is_favorite: false,
  name: `Dardjeeling Limited`,
  poster_image: `https://es31-server.appspot.com/wtw/static/film/poster/Dardjeeling_Limited.jpg`,
  preview_image: `https://es31-server.appspot.com/wtw/static/film/preview/dardjeeling_limited.jpg`,
  preview_video_link: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  rating: 3.6,
  released: 2007,
  run_time: 91,
  scores_count: 165106,
  starring: [`Owen Wilson`, `Adrien Brody`, `Jason Schwartzman`],
  video_link: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`
};
