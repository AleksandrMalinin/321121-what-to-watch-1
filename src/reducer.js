import {films} from './mocks/films.js';
import {constants} from './constants.js';

const initialState = {
  moviesList: films,
  activeGenre: constants.DEFAULT_GENRE
};

const ActionCreators = {
  changeGenre: (genre) => {
    return {
      type: constants.ACTION_TYPE.changeFilterGenre,
      payload: genre
    };
  },

  getMoviesList: (genre) => {
    let movies = genre === initialState.activeGenre ? initialState.moviesList.slice(1) : initialState.moviesList.filter((film) => genre === film.genre);

    return {
      type: constants.ACTION_TYPE.getMoviesList,
      payload: movies
    };
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case constants.ACTION_TYPE.changeFilterGenre: return Object.assign({}, state, {
      activeGenre: action.payload
    });

    case constants.ACTION_TYPE.getMoviesList: return Object.assign({}, state, {
      moviesList: action.payload
    });
  }

  return state;
};

export {reducer, ActionCreators};
