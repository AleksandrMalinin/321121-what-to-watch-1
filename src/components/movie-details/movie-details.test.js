import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import {Provider} from 'react-redux';
import {BrowserRouter} from 'react-router-dom';
import {createStore} from 'redux';
import reducer from '../../reducer';
import {MovieDetails} from './movie-details.jsx';
import {films, defaultMovie} from '../../mocks/mocks.js';

const mockStore = createStore(reducer);

it(`MovieDetails correctly renders`, () => {
  const renderer = new ShallowRenderer();
  const tree = renderer
  .render(
      <Provider store={mockStore}>
        <BrowserRouter>
          <MovieDetails
            movie={defaultMovie}
            moviesAlike={films}
            onCommentsLoad={jest.fn()}
          />)
        </BrowserRouter>
      </Provider>
  );

  expect(tree).toMatchSnapshot();
});
