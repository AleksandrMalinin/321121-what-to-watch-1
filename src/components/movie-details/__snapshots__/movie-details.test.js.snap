// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails correctly renders 1`] = `
<Context.Provider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <BrowserRouter>
    <MovieDetails
      handleCommentsLoad={[MockFunction]}
      movie={
        Object {
          "background_color": "#AD9F8B",
          "background_image": "https://es31-server.appspot.com/wtw/static/film/background/Dardjeeling_Limited.jpg",
          "description": "A year after their father's funeral, three brothers travel across India by train in an attempt to bond with each other.",
          "director": "Wes Anderson",
          "genre": "Adventure",
          "id": 1,
          "is_favorite": false,
          "name": "Dardjeeling Limited",
          "poster_image": "https://es31-server.appspot.com/wtw/static/film/poster/Dardjeeling_Limited.jpg",
          "preview_image": "https://es31-server.appspot.com/wtw/static/film/preview/dardjeeling_limited.jpg",
          "preview_video_link": "https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4",
          "rating": 3.6,
          "released": 2007,
          "run_time": 91,
          "scores_count": 165106,
          "starring": Array [
            "Owen Wilson",
            "Adrien Brody",
            "Jason Schwartzman",
          ],
          "video_link": "http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4",
        }
      }
      moviesAlike={
        Array [
          Object {
            "genre": "Comedies",
            "name": "Fantastic Beasts: The Crimes of Grindelwald",
            "preview_image": "img/fantastic-beasts-the-crimes-of-grindelwald.jpg",
            "preview_video_link": "https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4",
          },
          Object {
            "genre": "Crime",
            "name": "Bohemian Rhapsody",
            "preview_image": "img/bohemian-rhapsody.jpg",
            "preview_video_link": "https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4",
          },
          Object {
            "genre": "Documentary",
            "name": "Macbeth",
            "preview_image": "img/macbeth.jpg",
            "preview_video_link": "https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4",
          },
          Object {
            "genre": "Dramas",
            "name": "Aviator",
            "preview_image": "img/aviator.jpg",
            "preview_video_link": "https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4",
          },
          Object {
            "genre": "Horror",
            "name": "We need to talk about Kevin",
            "preview_image": "img/we-need-to-talk-about-kevin.jpg",
            "preview_video_link": "https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4",
          },
        ]
      }
    />
    )
  </BrowserRouter>
</Context.Provider>
`;
